src/util/src/WaveformUtil.cc

```diff
@@ -0,0 +1,251 @@
+  /*
+  Calculates the total number of threshold crossings, time over threshold, and voltage over threshold
+  */
+  int nCrossings = 0;
+  double timeOverThreshold = 0;
+  double voltageOverThreshold = 0;
+
+  bool crossed = false;
+  // Scan over the entire waveform
+  for (int i = 0; i < waveform.size(); i++) {
+    double voltage = waveform[i];
+
+    // If we crossed below threshold
+    if (voltage < voltageThreshold) {
+      // Not already below thresh, count the crossing
+      if (!crossed) {
+        nCrossings += 1;
+      }
+      // Count the time over threshold, mark that we crossed
+      timeOverThreshold += timeStep;
+      voltageOverThreshold += voltage;
+      crossed = true;
+    }
+    // If we are above threshold
+    if (voltage > voltageThreshold) {
+      crossed = false;
+    }
+  }
+  return std::make_tuple(nCrossings, timeOverThreshold, voltageOverThreshold);
+}
+
+double CalculateTimeCFD(const std::vector<double>& waveform, std::pair<int, double> peak, double constFrac,
+                        int lookBack, double timeStep) {
+  /*
+  Apply constant-fraction discriminator for a given peak
+  */
+  double voltageThreshold = constFrac * peak.second;
+  int time = GetThresholdCrossingBeforePeak(waveform, peak.first, voltageThreshold, lookBack, timeStep);
+  // Linearly interpolate threshold crossing time
+  double deltav = waveform.at(time + 1) - waveform.at(time);
+  double dt = timeStep / deltav * (voltageThreshold - waveform.at(time));
+  return time * timeStep + dt;
+}
+
+double IntegratePeak(const std::vector<double>& waveform, int peakSample, int intWindowLow, int intWindowHigh,
+                     double timeStep, double termOhms) {
+  /*
+  Integrate the digitized waveform around the given peak to calculate charge
+  */
+  double charge = 0;
+  int windowStart = peakSample - intWindowLow;
+  int windowEnd = peakSample + intWindowHigh;
+
+  if (windowStart >= waveform.size()) {
+    charge = INVALID;  // Invalid value for bad waveforms
+    return charge;
+  }
+
+  // Make sure not to integrate past the end of the waveform
+  windowEnd = (windowEnd > waveform.size()) ? waveform.size() : windowEnd;
+  // Make sure not to integrate before the waveform starts
+  windowStart = (windowStart < 0) ? 0 : windowStart;
+
+  for (int i = windowStart; i < windowEnd; i++) {
+    double voltage = waveform[i];
+    charge += VoltagetoCharge(voltage, timeStep, termOhms);  // in pC
+  }
+  return charge;
+}
+
+double IntegrateSliding(const std::vector<double>& waveform, int slidingWindow, double chargeThresh, double timeStep,
+                        double termOhms) {
+  /*
+  Integrate the digitized waveform over sliding windows to calculate a total charge
+  */
+  double total_charge = 0;
+  int nsliding = int(waveform.size() / slidingWindow);
+
+  for (int i = 0; i < nsliding; i++) {
+    double charge = 0;
+    int sample_start = i * slidingWindow;
+    int sample_end = (i + 1) * slidingWindow;
```

Comment:
I think there may be an edge case where sample_end  == waveform.size()?

Reply:
The subsequent loop is `for (int j = sample_start; j < sample_end; j++)`, which is fine if `sample_end == waveform.size()`, no?

Reply:
Yep, you're right.